File: android/app/src/main/java/com/example/pokemonsample/data/repository/PokemonRepositoryImpl.kt
package com.example.pokemonsample.data.repository

import com.example.pokemonsample.data.mapper.toDomain
import com.example.pokemonsample.data.remote.api.PokeApiService
import com.example.pokemonsample.domain.entity.Pokemon
import com.example.pokemonsample.domain.entity.Result
import com.example.pokemonsample.domain.repository.PokemonRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Dispatcher
import okio.IOException
import javax.inject.Inject
import javax.inject.Singleton

/**
 * PokemonRepository の実装クラス
 */
@Singleton
class PokemonRepositoryImpl @Inject constructor(
    private val apiService: PokeApiService
) : PokemonRepository {
    override suspend fun getPokemonList(limit: Int, offset: Int): Result<List<Pokemon>> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiService.getPokemonList(limit, offset)
                val pokemonList = response.results.map { it.toDomain() }
                Result.Success(pokemonList)
            } catch (e: IOException) {
                Result.Error(
                    exception = e,
                    message = "ネットワークエラーが発生しました"
                )
            } catch (e: Exception) {
                Result.Error(
                    exception = e,
                    message = "予期しないエラーが発生しました"
                )
            }
        }
    }

    override suspend fun searchPokemon(query: String): Result<List<Pokemon>> {
        // とりあえず空リストを返す
        return Result.Success(emptyList())
    }
}

File: ios/PokemonSample/Data/Repository/PokemonRepositoryImpl.swift
//
//  PokemonRepositoryImpl.swift
//  PokemonSample
//
//  Created by ponro on 2026/02/05.
//

import Foundation

/// ポケモンリポジトリの実装
class PokemonRepositoryImpl: PokemonRepository {
    private let apiService: PokeAPIServiceProtocol
    
    init(apiService: PokeAPIServiceProtocol) {
        self.apiService = apiService
    }
    
    func getPokemonList(limit: Int, offset: Int) async -> Result<(pokemons: [Pokemon], hasMore: Bool)> {
        do {
            let response = try await apiService.getPokemonList(limit: limit, offset: offset)
            let pokemons = response.results.map { $0.toDomain() }
            let hasMore = response.next != nil
            return .success((pokemons: pokemons, hasMore: hasMore))
        } catch let error as APIError {
            return .failure(error)
        } catch {
            return .failure(APIError.networkError)
        }
    }
    
    func getPokemonDetail(id: Int) async -> Result<PokemonDetail> {
        do {
            let response = try await apiService.getPokemonDetail(id: id)
            let detail = response.toDomain()
            return .success(detail)
        } catch let error as APIError {
            return .failure(error)
        } catch {
            return .failure(APIError.networkError)
        }
    }
    
    func getPokemonDetailByName(name: String) async -> Result<PokemonDetail> {
        do {
            let response = try await apiService.getPokemonDetailByName(name: name)
            let detail = response.toDomain()
            return .success(detail)
        } catch let error as APIError {
            return .failure(error)
        } catch {
            return .failure(APIError.networkError)
        }
    }
}
